include_physics/ideal_MHD.h

```diff
@@ -388,12 +388,177 @@ inline void init_fields(uint ix, uint iy, uint iz, global REAL* u) {
+//      flux[Field_By] = 0;
+        flux[Field_Bz] = u[Field_uy] * u[Field_Bz] - u[Field_uz]*u[Field_By];
+        return;
+}
+
+inline void calc_flux_h(REAL *u, REAL *flux){
+        REAL BB = u[Field_Bx]*u[Field_Bx] + u[Field_By]*u[Field_By] + u[Field_Bz]*u[Field_Bz];
+        flux[Field_rho] = u[Field_rho_uz];
+        flux[Field_rho_ux] = u[Field_rho_uz] * u[Field_ux] - u[Field_Bz]*u[Field_Bx];
+        flux[Field_rho_uy] = u[Field_rho_uz] * u[Field_uy] - u[Field_Bz]*u[Field_By];
+        flux[Field_rho_uz] = u[Field_rho_uz] * u[Field_uz] + u[Field_p] + 0.5 * BB - u[Field_Bz]*u[Field_Bz];
+        flux[Field_E] = u[Field_uz]*(u[Field_E] + u[Field_p] + 0.5*BB)-u[Field_Bz]*(u[Field_Bx]*u[Field_ux] + u[Field_By]*u[Field_uy] + u[Field_Bz]*u[Field_uz]);
+        flux[Field_Bx] = u[Field_uz] * u[Field_Bx]- u[Field_ux]*u[Field_Bz];
+        flux[Field_By] = u[Field_uz] * u[Field_By]- u[Field_uy]*u[Field_Bz];
+//      flux[Field_Bz] = 0;
+        return;
+}
+
+
+
+inline void calc_num_flux(REAL al, REAL ar, REAL *ul, REAL *ur, REAL *fluxl, REAL *fluxr, REAL *num_flux){
+        int i;
+        if(0 < al){
+                for (i = 0; i < NUM_CONSERVED_VARS; i++)
+                        num_flux[i] = fluxl[i];
+        } else if(0 < ar)
+                for (i = 0; i < NUM_CONSERVED_VARS; i++)
+                        num_flux[i] = (ar*fluxl[i]-al*fluxr[i])/(ar-al) + ar*al*(ur[i]-ul[i])/(ar-al);
+        else
+                for (i = 0; i < NUM_CONSERVED_VARS; i++)
+                        num_flux[i] = fluxr[i];
+
+}
+
+#endif
+
 inline void add_surface_terms(REAL time, uint ix, uint iy, uint iz, global REAL *u, REAL *du_dt) {
 
   // For periodic boundary conditions and a single block, no surface term has to be used.
 #ifndef USE_PERIODIC
-  #error "Error in ideal_MHD.cl: Surface terms not implemented for nonperiodic boundaries!"
-#endif // USE_PERIODIC*/
+  int i;
+  REAL um[NUM_TOTAL_VARS] = {0.0};
+  get_field(ix, iy, iz, 0, 0, 0, u, um); 
+  REAL4 b_bound = b_boundary(ix, iy, iz, time);
+  REAL4 u_bound = u_boundary(ix, iy, iz, time);
+  REAL rho_bound = (REAL)1;
+  REAL p_bound = 1;
+  REAL E_bound = compute_energy(p_bound, rho_bound, u_bound.x, u_bound.y, u_bound.z, b_bound.x, b_bound.y, b_bound.z);
+
+
+  REAL ub[NUM_TOTAL_VARS] = {0.0};
+  ub[Field_ux] = u_bound.x;
+  ub[Field_uy] = u_bound.y;
+  ub[Field_uz] = u_bound.z;
+  ub[Field_rho_ux] = u_bound.x * rho_bound;
+  ub[Field_rho_uy] = u_bound.y * rho_bound;
+  ub[Field_rho_uz] = u_bound.z * rho_bound;
+  ub[Field_rho] = rho_bound;
+  ub[Field_p] = p_bound;
+  ub[Field_E] = E_bound;
+  ub[Field_Bx] = b_bound.x;
+  ub[Field_By] = b_bound.y;
+  ub[Field_Bz] = b_bound.z;
+
+  REAL flux[NUM_CONSERVED_VARS] = {0.0};
+  REAL fluxm[NUM_CONSERVED_VARS] = {0.0};
+  REAL fluxb[NUM_CONSERVED_VARS] = {0.0};
+
+  REAL BBSQb = pow(ub[Field_Bx], 2) + pow(ub[Field_By], 2) + pow(ub[Field_Bz], 2);
+  REAL BBSQm = pow(um[Field_Bx], 2) + pow(um[Field_By], 2) + pow(um[Field_Bz], 2);
```

Comment:
Using simple products instead of `pow(..., 2)` results in a speedup of 15 % on my GPU.